gr-blocks/lib/correctiq_auto_impl.cc

```diff
@@ -0,0 +1,179 @@
+/* -*- c++ -*- */
+/*
+ * Copyright 2020 Free Software Foundation, Inc.
+ *
+ * This file is part of GNU Radio
+ *
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ *
+ */
+
+#ifdef HAVE_CONFIG_H
+#include "config.h"
+#endif
+
+#include "correctiq_auto_impl.h"
+#include <gnuradio/io_signature.h>
+#include <boost/format.hpp>
+
+namespace gr {
+namespace blocks {
+
+correctiq_auto::sptr
+correctiq_auto::make(double samp_rate, double freq, float gain, float sync_window)
+{
+    return gnuradio::get_initial_sptr(
+        new correctiq_auto_impl(samp_rate, freq, gain, sync_window));
+}
+
+/*
+ * The private constructor
+ */
+correctiq_auto_impl::correctiq_auto_impl(double samp_rate,
+                                         double freq,
+                                         float gain,
+                                         float sync_window)
+    : gr::sync_block("correctiq_auto",
+                     gr::io_signature::make(1, 1, sizeof(gr_complex)),
+                     gr::io_signature::make(1, 1, sizeof(gr_complex)))
+{
+    avg_real = 0.0;
+    avg_img = 0.0;
+
+    d_samp_rate = samp_rate;
+    d_freq = freq;
+    d_gain = gain;
+    d_sync_window = sync_window;
+
+    synchronized = false;
+
+    GR_LOG_INFO(d_logger, "Auto Synchronizing...");
+
+    d_max_syncSamples = (long)(d_samp_rate * (double)sync_window);
+
+    message_port_register_in(pmt::mp("rsync"));
+    set_msg_handler(pmt::mp("rsync"),
+                    [this](pmt::pmt_t msg) { this->handle_resync(msg); });
+
+    message_port_register_out(pmt::mp("sync_start"));
+    message_port_register_out(pmt::mp("offsets"));
+}
+
+void correctiq_auto_impl::send_sync_values()
+{
+    pmt::pmt_t meta = pmt::make_dict();
+
+    meta = pmt::dict_add(meta, pmt::mp("real"), pmt::mp(avg_real));
```

Comment:
Just a warning: `pmt::mp` calls `pmt::symbol` here, and that's relatively expensive (yes, that's stupid.).

One solution I've found for that is to have a static variable in the function you use the symbol, so that the symbol construction only happens once.

For now, this is definitely not a blocker. Especially since `send_sync_value` is not expected to be called extremely often.